README.md
# About 

SDK for WinnerMicro W800/W801

For W806, please use [wm-sdk-w806](https://github.com/IOsetting/wm-sdk-w806)

# File Structure

```
wm-sdk-w80x
├── app         # User application code
├── bin         
├── demo        # Demos
├── doc         # API documents
├── include
├── ld          # Link scripts
├── lib         # Libraries release in binary
├── Makefile
├── platform    # SDK source code
├── README.md
├── src
└── tools       # Utilities
```

# For Linux Users

## Download Toolchain
I upload the toolchain to github, you can download it from: https://github.com/cjacker/wm_sdk_w80x/releases/download/init/csky-elfabiv2-tools-x86_64-minilibc-20210423.tar.gz

Or download it from https://occ.t-head.cn/community/download.

## Installation

Extract the toolchains to proper folder -- be careful that the tar ball use `./` as top level path, move it to a seperate folder or specify a target folder for the uncompressing.
```bash
mkdir csky-elfabiv2-tools-x86_64-minilibc-20210423
tar xvf csky-elfabiv2-tools-x86_64-minilibc-20210423.tar.gz  -C csky-elfabiv2-tools-x86_64-minilibc-20210423/
```
Then move it to somewhere, e.g. /opt/toolchains, set it read-only to normal users
```bash
cd /opt/toolchains/
sudo mv ~/Download/csky-elfabiv2-tools-x86_64-minilibc-20210423/ .
sudo chown -R root:root csky-elfabiv2-tools-x86_64-minilibc-20210423/
```
You don't need to add it to the system PATH variable.

## Building

Checkout this project
```bash
git clone https://github.com/cjacker/wm_sdk_w80x.git
```

Then build the project
```bash
cd wm-sdk-w80x
make
```

## Configuration
Run menuconfig
```bash
cd wm-sdk-w80x
make menuconfig
```

## Help
```bash
make help
```

## Download To Development Board

Connect the development board to PC, get the USB port name by commands `dmesg`, `lsusb` and `ls /dev/tty*`.

Run menuconfig to set the download port
```bash
cd wm-sdk-w80x
make menuconfig
```
In menuconfig, navigate to `Download Configuration`
* Download port: Input the USB port name, e.g. `ttyUSB0`;
* Down rate: Input the UART baud rate, the higher rate the higher download speed. The availabe options are `115200`, `460800`, `921600`, `1000000` and `2000000`.

Then save and exit menuconfig, download the hex file to development board
```bash
make flash
```

Press the `Reset` key to start the downloading. If previously downloaded hex was built with `USE_UART0_AUTO_DL` enabled, the board will start downloading automatically.
```
build finished!
connecting serial...
serial connected.
wait serial sync.........         <--- Press the Reset key here
please manually reset the device. <--- (Or here)
.....
serial sync sucess.
mac CC-CC-CC-CC-CC-CC.
start download.
0% [###] 100%
download completed.
reset command has been sent.
```
When downloding finishes, the board will be reset automatically to run the new program. In case the auto-reset fails, you need to press the reset key manually to make it run.

### More Download Options

Show serial ports
```bash
make list
```
Build, download and start serial monitor
```bash
make run
```
Start serial monitor only
```bash
make monitor
```



doc/ChangeLog.txt
 ==========================================================
|	CHANGELOG: WinnerMicro Software Development Kit	   |
 ==========================================================
W800 SDK v1.00.04 | 2021/10/30
==========================
*********
SDK功能
********* 
1.驱动相关
  *低速SPI做主时，默认IO复用为PB2,PB3,PB4,PB5（与DEMO保持一致）
  *增加UART5功能
  *修复使用40M分频时32K不校准导致不准的问题，RTC使用40M分频时精度基本没有问题
  *修复LCD功能不工作的问题
  *增加ADC获取供电电压的功能
  *增加Flash的OTP操作，增加读取Unique ID的操作
  *增加touch功能
  *修改Flash驱动部分读取操作未加保护问题
  *修复去使能某个IO中断使能位时将整个IO中断也关闭的问题
  
2.DEMO相关
  *增加获取供电电压DEMO
  *增加PMU demo选择时钟使用

3.调试相关
  *增加打印口可设置为UART0/UART1的可配操作（默认使用UART0）

4.配网相关 
  *优改联盛德一键配网的内存使用
  *增加TLS_CONFIG_BLE_WIFI_ONESHOT宏定义

5.编译打包相关：
  *清理部分无效代码
  *修复PPP_SUPPORT宏定义打开，编译报错的问题
  *修复TLS_CONFIG_HOSTIF宏关闭，RMMS编译报错的问题
  *wm_tool.c打包工具更新，解决当升级文件小于1024Byte时打包出的文件无法升级成功的问题
  *增加CKLINK调试下载使用的flash驱动(tools/w800/utilities/W800_Flash_Tool.elf)

6.蓝牙相关
  *修复蓝牙wm_ble_client_demo_on_mtu设置MTU时不设限的问题

7.Wi-Fi相关
  *修复ssid为空时，tls_sys.c文件里自动联网依然触发问题
  *修复apssid为空时，依然可以创建热点成功的问题
  *优化Wi-Fi联网
  *优化联网低功耗
  *修复热点模式Beacon帧发送问题

8.AT相关
  *UART作为AT指令传输通道时，可指定为除UART0外的任意一个UART。
  *修复STA模式固定IP时，AT指令第二次加网，无法查询网络状态的问题
  
9.增加单独支持生产测试的AT指令(src/app/factorycmd目录下）
  *支持发送测试
  *支持接收测试
  *支持频偏补偿(AT+FREQ）
  *支持是否校准查询(AT+&CALFIN)

W800 SDK v1.00.02 | 2021/5/13
==========================
*********
SDK功能
*********
1. 简化RAM分配调整，详见wm_ram_config.h（除了HSPI和WiFi必须使用静态内存外）。

2. 代码空间调整说明：
   1）调整ld文件里的I-SRAM的LENGTH值，如果LENGTH值加ORIGIN值已越界FLASH容量，则调整ORIGIN的初始值
   2） 依据是CDK编译工程，还是直接make来选择调整
      （1） CDK工程里，要调整代码空间大小，除了步骤1）之外，要调整tools\w800\uitilities\aft_build_project.sh文件里的
	  run_img_header和run_img_pos，run_img_pos要和ORIGIN的值一致；run_img_header为ORIGIN值减0x400
      （2） 命令行方式编译，除了步骤1）之外，要调整CONFIG_W800_IMAGE_HEADER和CONFIG_W800_RUN_ADDRESS的值，方法同（1）
   3） 由于代码空间变大，用户空间会变小，那么需要注意用户空间的位置，作为使用时局限。

3. 蓝牙：
  1）ble host切换为nimble，默认使用的蓝牙协议栈
  2）增加了ble多连接示例
  3）蓝牙控制器代码优化
  4）标准蓝牙部分：提供了br_edr lib文件：libbtcontroller_br_edr.a和libbthost_br_edr.a
       如果需要使用标准蓝牙，需要操作：
       （1） 把wm_config.h里的宏TLS_CONFIG_BR_EDR设置为CFG_ON
       （2）编译时：
           如果使用命令行编译，需要make menuconfig，在编译选项菜单里，去使能nimble的选项，或者修改.config文件，把NIMBLE_FTR去掉
           如果使用CDK编译，需要把DNIMBLE_FTR设置为0，在CDK的编译设置项里
       （3）相应的需要把libbtcontroller_br_edr.a和libbthost_br_edr.a改为libbtcontroller.a和libbthost.a
       （4）还需要单独编译一下src/app目录（命令行编译）
       （5） 因蓝牙增加了Code Size，需扩大gcc.ld文件里I-SRAM的空间

   5）蓝牙任务删除统一修改为通过句柄方式删除
   6）蓝牙host的代码统一开源，位置位于目录:src\bt

4. Wi-Fi优化：
   1） STA节能处理优化
   2） 软ap功能暂不支持wmm功能

5. 驱动问题处理：
   DMA：    修复DMA是用loop方式时，循环地址处理不正确导致的中断不产生的问题
   ADC：    DMA方式回调接口统一，DMA时使用内存为动态申请内存；ADC的回调函数统一处理
   GPIO：   修复GPIOB的中断关闭操作时，没有正确操作的问题
   FLASH：  修复Flash加锁操作的操作数异常
   PMU：    删除PMU中断处理里的重复的csi_kernel_intrpt_exit调用触发调度的操作
   7816：   增加7816接口demo
   SPI：    低速主SPI，当使用DMA方式传输时，所用内存为动态分配内存
   IO复用： 增加了W801的IO复用功能设置接口
   uart：   为了适配某些平台，增加的单个字符即回调的接口。

6. OS：
   1） 增加通过任务句柄删除任务的操作(wm_osal_rtos.c, tasks.c)
   2） 完善通过优先级删除任务的操作(wm_osal_rtos.c, tasks.c)
   3） 部分os的适配接口，增加了中断里的使用保护(wm_osal_rtos.c)

7. 网络部分修改：
    1）网络主机名改为：WinnerMicro_+mac地址末2字段的格式。(ethernetif.c)
	2）修复sys_arch.c里的queue delete函数，不再使用开关中断保护

8. wm_main.c里的主任务增加可删除操作，可通过打开宏开关，使得wm_main里的任务使用完毕后删除

9. 内存分配（wm_mem.c）的封装函数改为仅用信号量来管理

10. 应用修复：
   1）FATFS文件系统增加可重入保护处理，针对SDIO HOST部分对接的写SD卡时的非4字节对齐做了处理。
   2）SSL不再使用之前的版本，改为使用mbed TLS 2.7.10
   3）修复oneshot宏关闭，链接失败的问题

11. 清理部分编译告警


W800 SDK v1.00.01 | 2020/12/29
==========================
*********
SDK功能
*********
*ld调整：
     因增加BT功能，代码空间超过1MB，当前设定为1MB+64KB，相应的用户区要减少64KB
     用户区宏调整（wm_flash_map.h）：#define USER_ADDR_START    (0x81E0000UL)
     当前升级区针对超过1MB的情况，只能选择使用压缩的ota固件

*二级BOOT更新：
    1）发布SDK里添加secboot.bin文件生成img的操作，以免客户调整运行区起始地址时，因为secboot的头信息与其不匹配导致升级fls启动不了	 

*蓝牙：
     1）提供三种蓝牙的库，bt和ble都有的lib，单独标准bt的lib，单独ble的lib
        用户可根据实际需要配合wm_config.h里的TLS_CONFIG_BR_EDR和TLS_CONFIG_BLE的设置，把相关的lib更名为libbt.a。
     2）发布时默认使用ble都有的lib
	 3）优化蓝牙参数
	 4）增加蓝牙demo

*Wi-Fi：

demo/wm_demo.h
#ifndef __WM_DEMO_H__
#define __WM_DEMO_H__

#define DEMO_ON		1
#define DEMO_OFF	0

//demo console
#define DEMO_CONSOLE				DEMO_OFF

//socket demo
#define DEMO_STD_SOCKET_CLIENT	    (DEMO_OFF && DEMO_CONSOLE)
#define DEMO_STD_SOCKET_SERVER	    (DEMO_OFF && DEMO_CONSOLE)

#define DEMO_UDP                    (DEMO_OFF && DEMO_CONSOLE)

//connect demo
#define DEMO_CONNECT_NET			(DEMO_OFF && DEMO_CONSOLE)

#define DEMO_SOCKET_CLIENT_SERVER   (DEMO_OFF && DEMO_CONSOLE)

//apsta demo
#define DEMO_APSTA   		  		(DEMO_OFF && DEMO_CONSOLE)

//Uart rx&tx demo
#define DEMO_UARTx   		  		(DEMO_OFF && DEMO_CONSOLE)

//AP dmeo
#define DEMO_SOFT_AP   		  		(DEMO_OFF && DEMO_CONSOLE)

//WPS demo
#define DEMO_WPS					(DEMO_OFF && DEMO_CONSOLE && TLS_CONFIG_WPS && TLS_IEEE8021X_EAPOL)

//gpio demo
#define DEMO_GPIO					(DEMO_OFF && DEMO_CONSOLE)

//flash demo
#define DEMO_FLASH					(DEMO_OFF && DEMO_CONSOLE)

//master spi demo
#define DEMO_MASTER_SPI				(DEMO_OFF && DEMO_CONSOLE)

//Encryption&Decryption demo
#define DEMO_ENCRYPT				(DEMO_OFF && DEMO_CONSOLE)

//rsa demo
#define DEMO_RSA					(DEMO_OFF && DEMO_CONSOLE)

//rtc demo
#define DEMO_RTC					(DEMO_OFF && DEMO_CONSOLE)

//timer demo
#define DEMO_TIMER					(DEMO_OFF && DEMO_CONSOLE)

//i2c demo
#define DEMO_I2C					(DEMO_OFF && DEMO_CONSOLE)

//pwm demo
#define DEMO_PWM					(DEMO_OFF && DEMO_CONSOLE)

//pmu demo
#define DEMO_PMU					(DEMO_OFF && DEMO_CONSOLE)

//ntp demo
#define DEMO_NTP					(DEMO_OFF && DEMO_CONSOLE)

//i2s demo
#define DEMO_I2S					(DEMO_OFF && DEMO_CONSOLE)

//adc demo
#define DEMO_ADC					(DEMO_OFF && DEMO_CONSOLE)

//slave spi demo
#define DEMO_SLAVE_SPI				(DEMO_OFF && DEMO_CONSOLE)

//7816
#define DEMO_7816					(DEMO_OFF && DEMO_CONSOLE)

//touch sensor
#define DEMO_TOUCHSENSOR           (DEMO_OFF && DEMO_CONSOLE)

//ssl server demo
#define DEMO_SSL_SERVER				(DEMO_OFF && DEMO_CONSOLE)

//websocket demo
#define DEMO_WEBSOCKETS				(DEMO_OFF && DEMO_CONSOLE)

//http demo
#define DEMO_HTTP					(DEMO_OFF && DEMO_CONSOLE)

#define DEMO_SCAN					(DEMO_OFF && DEMO_CONSOLE)
#define DEMO_HTTPS                  (DEMO_OFF && DEMO_CONSOLE)
#define DEMO_MQTT                   (DEMO_OFF && DEMO_CONSOLE)
#define DEMO_SDIO_HOST              (DEMO_OFF && DEMO_CONSOLE)
#define DEMO_FATFS                  (DEMO_OFF && DEMO_CONSOLE)
#define DEMO_MBEDTLS                (DEMO_OFF && DEMO_CONSOLE)


#define DEMO_DSP					(DEMO_OFF && DEMO_CONSOLE)

#define DEMO_BT                     (DEMO_OFF && DEMO_CONSOLE)

////////////////////////////////////////////////////////////////

// user prio 32 - 60
#define  DEMO_TASK_PRIO			                32
#define  DEMO_SOCKET_C_TASK_PRIO	            (DEMO_TASK_PRIO + 1)
#define  DEMO_SOCKET_S_TASK_PRIO	            (DEMO_SOCKET_C_TASK_PRIO + 1)
#define  DEMO_SOCKET_RECEIVE_TASK_PRIO	        (DEMO_SOCKET_S_TASK_PRIO + 1)
#define  DEMO_SOCKET_S_SEND_TASK_PRIO    	    (DEMO_SOCKET_RECEIVE_TASK_PRIO + 1)
#define  DEMO_SOCKET_S_RECEIVE_TASK_PRIO    	(DEMO_SOCKET_S_SEND_TASK_PRIO + 1)
#define  DEMO_UDP_TASK_PRIO                     (DEMO_SOCKET_S_RECEIVE_TASK_PRIO + 1)
#define  DEMO_UDP_RECEIVE_TASK_PRIO             (DEMO_UDP_TASK_PRIO + 1)
#define  DEMO_MCAST_TASK_PRIO	                (DEMO_UDP_RECEIVE_TASK_PRIO + 1)
#define  DEMO_SOCK_FWUP_TASK_PRIO	            (DEMO_MCAST_TASK_PRIO + 1)
#define  DEMO_UART_TASK_PRIO	                (DEMO_SOCK_FWUP_TASK_PRIO + 1)
#define  DEMO_SSL_SERVER_TASK_PRIO	            (DEMO_UART_TASK_PRIO + 1)
#define  DEMO_WEBSOCKETS_TASK_PRIO              (DEMO_SSL_SERVER_TASK_PRIO + 1)

#define DEMO_QUEUE_SIZE	32

#define DEMO_BUF_SIZE		TLS_UART_RX_BUF_SIZE


/*Message Definition used for Demo task*/
#define	DEMO_MSG_WJOIN_FAILD	            1
#define	DEMO_MSG_WJOIN_SUCCESS	            2
#define DEMO_MSG_SOCKET_RECEIVE_DATA		3
#define	DEMO_MSG_UART_RECEIVE_DATA			4
#define	DEMO_MSG_SOCKET_ERR					5
#define DEMO_MSG_SOCKET_CREATE				6
#define DEMO_MSG_SOCKET_TEST				7
#define DEMO_MSG_OPEN_UART					8
#define DEMO_MSG_UART_CMD_DATA				9

typedef struct demo_sys
{
    char *rxbuf;		   /*uart rx*/
    char *txbuf;		   /*uart tx*/
    u16 wptr;
    u16 rptr;
    u8 dummy1;		      /*not used*/
    u8 MsgNum;


    bool socket_ok;
    int socket_num;
    char *sock_rx;
    u16 sock_data_len;

    int recvlen;		/*socket rx length, used to debug*/
    int translen;		/*socket tx length, used to debug*/
    u8 dummy2; 			/*not used*/
} ST_Demo_Sys;

void CreateDemoTask(void);

extern int wm_printf(const char *fmt,...) ;

#endif


include/driver/wm_uart.h
/**
 * @file    wm_uart.h
 *
 * @brief   uart Driver Module
 *
 * @author  dave
 *
 * Copyright (c) 2015 Winner Microelectronics Co., Ltd.
 */
#ifndef WM_UART_H
#define WM_UART_H
#include "list.h"
//#include "wm_regs.h"
#include "wm_type_def.h"
#include "wm_osal.h"

#define TLS_UART_RX_BUF_SIZE   4096
#define TLS_UART_TX_BUF_SIZE   4096
#define WAKEUP_CHARS           256

#define MBOX_MSG_UART_RX       1
#define MBOX_MSG_UART_TX       2


/** baud rate definition */
#define UART_BAUDRATE_B600          600
#define UART_BAUDRATE_B1200	        1200
#define UART_BAUDRATE_B1800         1800
#define UART_BAUDRATE_B2400         2400
#define UART_BAUDRATE_B4800         4800
#define UART_BAUDRATE_B9600         9600
#define UART_BAUDRATE_B19200        19200
#define UART_BAUDRATE_B38400        38400
#define UART_BAUDRATE_B57600        57600
#define UART_BAUDRATE_B115200       115200
#define UART_BAUDRATE_B230400       230400
#define UART_BAUDRATE_B460800       460800
#define UART_BAUDRATE_B921600      921600
#define UART_BAUDRATE_B1000000      1000000
#define UART_BAUDRATE_B1250000      1250000
#define UART_BAUDRATE_B1500000      1500000
#define UART_BAUDRATE_B2000000      2000000

#define UART_RX_INT_FLAG (UIS_RX_FIFO | UIS_RX_FIFO_TIMEOUT | UIS_BREAK |\
        UIS_OVERRUN | UIS_FRM_ERR | UIS_PARITY_ERR)
#define UART_RX_ERR_INT_FLAG (UIS_BREAK | UIS_FRM_ERR | \
        UIS_PARITY_ERR)

#define UART_TX_INT_FLAG (UIS_TX_FIFO | UIS_TX_FIFO_EMPTY)

/** return count in buffer.  */
#define CIRC_CNT(head,tail,size) (((head) - (tail)) & ((size)-1))

/** Return space available, 0..size-1.  We always leave one free char
   as a completely full buffer has head == tail, which is the same as
   empty.  */
#define CIRC_SPACE(head,tail,size) CIRC_CNT((tail),((head)+1),(size))

/** Return count up to the end of the buffer.  Carefully avoid
   accessing head and tail more than once, so they can change
   underneath us without returning inconsistent results.  */
#define CIRC_CNT_TO_END(head,tail,size) \
	({int end = (size) - (tail); \
	  int n = ((head) + end) & ((size)-1); \
	  n < end ? n : end;})

/** Return space available up to the end of the buffer.  */
#define CIRC_SPACE_TO_END(head,tail,size) \
	({int end = (size) - 1 - (head); \
	  int n = (end + (tail)) & ((size)-1); \
	  n <= end ? n : end+1;})

#define CIRC_SPACE_TO_END_FULL(head,tail,size) \
	({int end = (size) - 1 - (head); \
	  int n = (end + (tail)) & ((size)-1); \
	  n < end ? n : end+1;})

#define uart_circ_empty(circ)		((circ)->head == (circ)->tail)
#define uart_circ_chars_pending(circ)	\
	(CIRC_CNT((circ)->head, (circ)->tail, TLS_UART_TX_BUF_SIZE))

/**
 * @struct tls_uart_baud_rate    baudrate define
 */
struct tls_uart_baud_rate
{
    u32 baud_rate;
    u16 ubdiv;
    u16 ubdiv_frac;
};


/**
 * @enum    uart number enum
 */
enum
{
    TLS_UART_0 = 0,
    TLS_UART_1 = 1,
    TLS_UART_2 = 2,
    TLS_UART_3 = 3,
    TLS_UART_4 = 4,
    TLS_UART_5 = 5,    
    TLS_UART_MAX = 6,
};


/**
 * @typedef enum TLS_UART_PMODE    Parity Mode
 */
typedef enum TLS_UART_PMODE
{
    TLS_UART_PMODE_DISABLED = 0,    /**< No Parity */
    TLS_UART_PMODE_ODD = 1,     /**< Odd Parity */
    TLS_UART_PMODE_EVEN = 2,    /**< Even Parity */
    TLS_UART_PMODE_MARK = 3,    /**< The parity bit is always 1. */
    TLS_UART_PMODE_SPACE = 4,   /**< The parity bit is always 0. */
} TLS_UART_PMODE_T;

/**
 * @typedef enum TLS_UART_CHSIZE    Character Size
 */
typedef enum TLS_UART_CHSIZE
{
    TLS_UART_CHSIZE_5BIT = (0x00 << 0), /**< Character size: 5 bit */
    TLS_UART_CHSIZE_6BIT = (0x01 << 0), /**< Character size: 6 bit */
    TLS_UART_CHSIZE_7BIT = (0x02 << 0), /**< Character size: 7 bit */
    TLS_UART_CHSIZE_8BIT = (0x03 << 0), /**< Character size: 8 bit */
} TLS_UART_CHSIZE_T;

/**
 * @typedef enum TLS_UART_FLOW_CTRL_MODE    flow control mode
 */
typedef enum TLS_UART_FLOW_CTRL_MODE
{
    TLS_UART_FLOW_CTRL_NONE,
    TLS_UART_FLOW_CTRL_HARDWARE,
} TLS_UART_FLOW_CTRL_MODE_T;

/**
 * @typedef enum TLS_UART_RX_FLOW_CTRL_FLAG    flow control rx flag
 */
typedef enum TLS_UART_RX_FLOW_CTRL_FLAG
{
    TLS_UART_RX_DISABLE,
    TLS_UART_RX_ENABLE,
} TLS_UART_RX_FLOW_CTRL_FLAG_T;

/**
 * @typedef enum TLS_UART_STOPBITS
 */
typedef enum TLS_UART_STOPBITS
{
    TLS_UART_ONE_STOPBITS,
    TLS_UART_TWO_STOPBITS,
} TLS_UART_STOPBITS_T;


/**
 * @typedef enum TLS_UART_STATUS
 */
typedef enum TLS_UART_STATUS
{
    TLS_UART_STATUS_OK,
    TLS_UART_STATUS_ERROR,
} TLS_UART_STATUS_T;


/**
 * @typedef enum TLS_UART_MODE   operation mode
 */
typedef enum TLS_UART_MODE
{
    TLS_UART_MODE_POLL,         /**< uart operation mode: poll */
    TLS_UART_MODE_INT,          /**< uart operation mode: interrupt mode */
} TLS_UART_MODE_T;

/**
 * @struct tls_uart_icount
 */
struct tls_uart_icount
{
    u32 cts;
    u32 dsr;
    u32 rng;
    u32 dcd;
    u32 rx;
    u32 tx;
    u32 frame;
    u32 overrun;
    u32 parity;
    u32 brk;
    u32 buf_overrun;
};



/**
 * @typedef struct tls_uart_options
 */
typedef struct tls_uart_options
{
    u32 baudrate;    /**< Set baud rate of the UART */

    TLS_UART_CHSIZE_T charlength;   /**< Number of bits to transmit as a character (5 to 8). */

    TLS_UART_PMODE_T paritytype;    /**< Parity type */

    TLS_UART_FLOW_CTRL_MODE_T flow_ctrl;    /**< Flow control type */

    TLS_UART_STOPBITS_T stopbits;    /**< Number of stop bits */

} tls_uart_options_t;


/**
 * @typedef struct tls_uart_circ_buf
 */
typedef struct tls_uart_circ_buf
{
volatile    u8 *buf;
volatile   u32 head;
volatile   u32 tail;
} tls_uart_circ_buf_t;

#if TLS_CONFIG_CMD_NET_USE_LIST_FTR
/**
 * @typedef struct tls_uart_net_buf
 */
typedef struct tls_uart_net_buf
{
    struct dl_list list;
    char *buf;
	void *pbuf;
    u16 buflen;
    u16 offset;
} tls_uart_net_buf_t;

typedef struct tls_uart_net_msg
{
    struct dl_list tx_msg_pending_list;
} tls_uart_net_msg_t;
#endif


/**
 * @typedef struct TLS_UART_REGS
 */
typedef struct TLS_UART_REGS
{
    u32 UR_LC;                       /**< line control register */
    u32 UR_FC;                       /**<  flow control register */
    u32 UR_DMAC;                  /**< dma control register */
    u32 UR_FIFOC;                  /**< fifo control register */
    u32 UR_BD;                       /**< baud rate register */
    u32 UR_INTM;                   /**< interrupt mask register */
    u32 UR_INTS;                    /**< interrupt source register */
    u32 UR_FIFOS;                  /**< fifo status register */
    u32 UR_TXW;                    /**< tx windows register */
    u32 UR_RES0;
    u32 UR_RES1;
    u32 UR_RES2;
    u32 UR_RXW;                     /**< rx windows register */
} TLS_UART_REGS_T;


/**
 * @typedef struct tls_uart_port
 */
typedef struct tls_uart_port
{
    u32 uart_no;                    /**< uart number: 0 or 1 */

    u32 uart_irq_no;             /**< uart interrupt number */

    u32 plus_char_cnt;

    TLS_UART_MODE_T uart_mode;      /**< uart work mode: interrupt mode or poll mode */

    struct tls_uart_options opts;       /**< uart config parameters */

    int fcStatus;                           /**< flow ctrl status,0 closed ,1 opened */

    enum TLS_UART_RX_FLOW_CTRL_FLAG rxstatus;

    u32 tx_fifofull;                    /**< uart tx fifo trigger level */

    TLS_UART_REGS_T volatile *regs;     /**< uart registers struct pointer */

    struct tls_uart_icount icount;          /**< uart statistics information */

    struct tls_uart_circ_buf recv;          /**< uart ring buffer */

// struct tls_uart_circ_buf xmit;

    struct dl_list tx_msg_pending_list;

    struct dl_list tx_msg_to_be_freed_list;

    u8 hw_stopped;

    tls_os_sem_t *tx_sem;

    char *buf_ptr;

    u16 buf_len;

    s16(*rx_callback) (u16 len, void* priv_data);

    s16(*tx_callback) (struct tls_uart_port * port);
    s16(*tx_sent_callback) (struct tls_uart_port * port);

    bool tx_dma_on;
	bool rx_dma_on;

	void *priv_data;
} tls_uart_port_t;

/**
 * @typedef struct tls_uart_tx_msg
 */
typedef struct tls_uart_tx_msg
{
    struct dl_list list;
    char *buf;
    u16 buflen;
    u16 offset;
    void (*finish_callback) (void *arg);
    void *callback_arg;
} tls_uart_tx_msg_t;

/**
 * @defgroup Driver_APIs Driver APIs
 * @brief Driver APIs
 */

/**
 * @addtogroup Driver_APIs
 * @{
 */

/**
 * @defgroup UART_Driver_APIs UART Driver APIs
 * @brief UART driver APIs
 */

/**
 * @addtogroup UART_Driver_APIs
 * @{
 */

/**
 * @brief	This function is used to initial uart port.
 *
 * @param[in] uart_no: is the uart number.
 *	- \ref TLS_UART_0 TLS_UART_1 TLS_UART_2 TLS_UART_3 TLS_UART_4 TLS_UART_5
 * @param[in] opts: is the uart setting options,if this param is NULL,this function will use the default options.
 * @param[in] modeChoose:; choose uart2 mode or 7816 mode when uart_no is TLS_UART_2, 0 for uart2 mode and 1 for 7816 mode.
 *
 * @retval
 *	- \ref WM_SUCCESS
 *	- \ref WM_FAILED
 *
 * @note When the system is initialized, the function has been called, so users can not call the function.
 */
int tls_uart_port_init(u16 uart_no, tls_uart_options_t * opts, u8 modeChoose);


/**
 * @brief          This function is used to register uart rx interrupt.
 *
 * @param[in]      uart_no      TLS_UART_0 or TLS_UART_1
 * @param[in]      rx_callback  is the uart rx interrupt call back function.
 *
 * @return         None
 *
 * @note           None
 */
void tls_uart_rx_callback_register(u16 uart_no, s16(*rx_callback) (u16 len, void* user_data), void* user_data);

void tls_uart_rx_byte_callback_flag(u16 uart_no, u8 flag);

/**
 * @brief	This function is used to register uart tx interrupt.
 *
 * @param[in] uart_no: is the uart numer.
 * @param[in] callback: is the uart tx interrupt call back function.
 *
 * @retval
 */
void tls_uart_tx_callback_register(u16 uart_no, s16(*tx_callback) (struct tls_uart_port *port));


/**
 * @brief          This function is used to copy circular buffer data to user buffer.
 *
 * @param[in]      uart_no    is the uart numer
 * @param[in]      buf          is the user buffer
 * @param[in]      readsize   is the user read size
 *
 * @retval         copy data size
 *
 * @note           None
 */
int tls_uart_read(u16 uart_no, u8 * buf, u16 readsize);

/**
 * @brief          This function is used to check the available data in the cache buffer.
 *
 * @param[in]      uart_no    is the uart numer
 * @param[in]      readsize   is the user read size
 *
 * @retval         if the cache buffer size is greater or equals to readsize , then return readsize; otherwise return 0;
 *
 * @note           None
 */

int tls_uart_try_read(u16 uart_no, int32_t read_size);


/**
 * @brief          This function is used to transfer data synchronously.
 *
 * @param[in]      uart_no      is the uart number
 * @param[in]      buf            is a buf for saving user data
 * @param[in]      writesize    is the user data length
 *
 * @retval         WM_SUCCESS    tx success
 * @retval         WM_FAILED       tx failed
 *
 * @note           None
 */
int tls_uart_write(u16 uart_no, char *buf, u16 writesize);


/**
 * @brief          This function is used to transfer data with DMA.
 *
 * @param[in]      buf                is a buf for saving user data
 * @param[in]      writesize        is the user data length
 * @param[in]      cmpl_callback  function point,when the transfer is completed, the function will be called.
 *
 * @retval         WM_SUCCESS    success
 * @retval         WM_FAILED       failed
 *
 * @note           Only uart1 support DMA transfer.
 */
int tls_uart_dma_write(char *buf, u16 writesize, void (*cmpl_callback) (void *p), u16 uart_no);


/**
 * @brief          This function is used to set uart parity.
 *
 * @param[in]      uart_no      is the uart number
 * @param[in]      paritytype   is a parity type defined in TLS_UART_PMODE_T
 *
 * @retval         WM_SUCCESS	if setting success
 * @retval         WM_FAILED	 	if setting fail
 *
 * @note           None
 */
int tls_uart_set_parity(u16 uart_no, TLS_UART_PMODE_T paritytype);


/**
 * @brief          This function is used to set uart baudrate.
 *
 * @param[in]      uart_no     is the uart number
 * @param[in]      baudrate    is the baudrate user want used,the unit is HZ.
 *
 * @retval         WM_SUCCESS	if setting success
 * @retval         WM_FAILED 	if setting fail
 *
 * @note           None
 */
int tls_uart_set_baud_rate(u16 uart_no, u32 baudrate);

/**
 * @brief          This function is used to set uart stop bits.
 *
 * @param[in]      uart_no     is the uart number
 * @param[in]      stopbits    is a stop bit type defined in TLS_UART_STOPBITS_T
 *
 * @retval         WM_SUCCESS	if setting success
 * @retval         WM_FAILED	if setting fail
 *
 * @note           None
 */
int tls_uart_set_stop_bits(u16 uart_no, TLS_UART_STOPBITS_T stopbits);

/**
 * @}
 */

/**
 * @}
 */
void tls_uart_push(int uart_no, u8* data, int length);


/**
 * @brief          This function is used to transfer data asynchronously.
 *
 * @param[in]      uart_no      is the uart number
 * @param[in]      buf            is a buf for saving user data
 * @param[in]      writesize    is the user data length
 *
 * @retval         WM_SUCCESS    tx success
 * @retval         WM_FAILED       tx failed
 *
 * @note           None
 */

int tls_uart_write_async(u16 uart_no, char *buf, u16 writesize);

/**
 * @brief	This function is used to register uart tx sent callback function.
 *
 * @param[in] uart_no: is the uart numer.
 * @param[in] callback: is the uart tx sent out call back function.
 *
 * @retval
 */

void tls_uart_tx_sent_callback_register(u16 uart_no, s16(*tx_callback) (struct tls_uart_port *port));



#endif /* WM_UART_H */
